File: ios/ThreeDSecureExecutor.swift
//
//  ThreeDSecureExecutor.swift
//  expo-braintree
//
//  Standalone 3D Secure verification executor
//

import Braintree
import Foundation
import UIKit

class ThreeDSecureExecutor: NSObject, BasePaymentExecutor, BTThreeDSecureRequestDelegate {
  let args: BasePaymentArgs
  weak var listener: PaymentExecutorListener?
  private var threeDSecureClient: BTThreeDSecureClient?
  private weak var viewController: UIViewController?

  init(args: BasePaymentArgs, listener: PaymentExecutorListener?) {
    self.args = args
    self.listener = listener
    super.init()
  }

  func requestPayment(viewController: UIViewController) {
    self.viewController = viewController

    guard case .threeDSecure(let nonce, let amount, let options) = args.paymentMethod else {
      handleError(message: "Invalid payment method", localizedMessage: "Expected ThreeDSecure")
      return
    }

    guard let apiClient = BTAPIClient(authorization: args.clientToken) else {
      handleError(
        message: ERROR_TYPES.API_CLIENT_INITIALIZATION_ERROR.rawValue,
        localizedMessage: "Failed to initialize Braintree API client"
      )
      return
    }

    threeDSecureClient = BTThreeDSecureClient(apiClient: apiClient)

    guard let secureClient = threeDSecureClient else {
      handleError(
        message: ERROR_TYPES.API_CLIENT_INITIALIZATION_ERROR.rawValue,
        localizedMessage: "Failed to initialize 3D Secure client"
      )
      return
    }

    var requestOptions = options
    requestOptions["nonce"] = nonce
    requestOptions["amount"] = amount
    requestOptions["email"] = args.email

    let threeDSecureRequest = prepareThreeDSecureRequest(options: requestOptions)
    threeDSecureRequest.threeDSecureRequestDelegate = self

    secureClient.startPaymentFlow(threeDSecureRequest) { [weak self] (threeDSecureResult, error) in
      guard let self = self else { return }

      if let threeDSecureResult = threeDSecureResult,
         let tokenizedCard = threeDSecureResult.tokenizedCard {

        let threeDSecureInfo = tokenizedCard.threeDSecureInfo

        if !threeDSecureInfo.liabilityShiftPossible {
          self.handleError(
            message: ERROR_TYPES.THREE_D_SECURE_NOT_ABLE_TO_SHIFT_LIABILITY.rawValue,
            localizedMessage: "3D Secure liability shift not possible"
          )
          return
        }

        if !threeDSecureInfo.liabilityShifted {
          self.handleError(
            message: ERROR_TYPES.THREE_D_SECURE_LIABILITY_NOT_SHIFTED.rawValue,
            localizedMessage: "3D Secure liability not shifted"
          )
          return
        }

        let additionalData = prepareThreeDSecureNonceResult(nonce: threeDSecureResult) as? [String: Any]

        self.handleSuccess(
          nonce: tokenizedCard.nonce,
          amount: amount,
          currency: "USD",
          paymentType: "Card3DS",
          additionalData: additionalData
        )
      } else if let error = error {
        self.handleError(
          message: ERROR_TYPES.THREE_D_SECURE_AUTHENTICATION_FAILED.rawValue,
          localizedMessage: error.localizedDescription
        )
      } else {
        self.handleError(
          message: ERROR_TYPES.THREE_D_SECURE_VERIFICATION_FAILED.rawValue,
          localizedMessage: "3D Secure verification failed"
        )
      }
    }
  }

  func onLookupComplete(_ request: BTThreeDSecureRequest, lookupResult: BTThreeDSecureResult, next: @escaping () -> Void) {
    next()
  }

  func onResume() {
  }

  func onDestroy() {
    threeDSecureClient = nil
  }
}

File: android/src/main/java/com/expobraintree/ThreeDSecureExecutor.kt
package com.expobraintree

import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.fragment.app.FragmentActivity
import com.braintreepayments.api.threedsecure.ThreeDSecureClient
import com.braintreepayments.api.threedsecure.ThreeDSecureLauncher
import com.braintreepayments.api.threedsecure.ThreeDSecureLauncherCallback
import com.braintreepayments.api.threedsecure.ThreeDSecurePaymentAuthRequest
import com.braintreepayments.api.threedsecure.ThreeDSecurePaymentAuthResult
import com.braintreepayments.api.threedsecure.ThreeDSecureRequest
import com.braintreepayments.api.threedsecure.ThreeDSecureResult

class ThreeDSecureExecutor(
    private val context: Context,
    listener: PaymentExecutorListener,
    args: BasePaymentArgs
) : BasePaymentExecutor(args, listener), ThreeDSecureLauncherCallback {

    private val threeDSecureClient: ThreeDSecureClient
    private var threeDSecureLauncher: ThreeDSecureLauncher? = null

    init {
        Log.d(TAG, "[init] Creating ThreeDSecureClient, tokenPresent=${args.clientToken.isNotEmpty()}")
        threeDSecureClient = ThreeDSecureClient(
            context = context,
            authorization = args.clientToken
        )
    }

    override fun requestPayment(activity: FragmentActivity) {
        val paymentMethod = args.paymentMethod as? PaymentMethod.ThreeDSecure
        if (paymentMethod == null) {
            Log.e(TAG, "[requestPayment] Invalid payment method")
            handleError("Invalid payment method", "Expected ThreeDSecure")
            return
        }

        Log.d(TAG, "[requestPayment] Starting 3DS verification, amount=${paymentMethod.amount}")

        threeDSecureLauncher = ThreeDSecureLauncher(activity, this)

        val threeDSecureRequest = ThreeDSecureRequest(
            nonce = paymentMethod.nonce,
            amount = paymentMethod.amount,
            email = paymentMethod.email
        )

        Log.d(TAG, "[requestPayment] Creating 3DS payment auth request")

        threeDSecureClient.createPaymentAuthRequest(
            context = activity,
            request = threeDSecureRequest
        ) { paymentAuthRequest ->
            Log.d(TAG, "[requestPayment] createPaymentAuthRequest callback, type: ${paymentAuthRequest::class.simpleName}")

            when (paymentAuthRequest) {
                is ThreeDSecurePaymentAuthRequest.ReadyToLaunch -> {
                    Log.d(TAG, "[requestPayment] Ready to launch 3DS, launching...")
                    threeDSecureLauncher?.launch(paymentAuthRequest)
                }
                is ThreeDSecurePaymentAuthRequest.LaunchNotRequired -> {
                    val nonce = paymentAuthRequest.nonce
                    val info = nonce.threeDSecureInfo
                    Log.d(TAG, "[requestPayment] 3DS launch not required, liabilityShifted=${info.liabilityShifted}, liabilityShiftPossible=${info.liabilityShiftPossible}")
                    handleSuccess(
                        nonce = nonce.string,
                        amount = paymentMethod.amount,
                        currency = paymentMethod.currency,
                        paymentType = "Card3DS",
                        threeDSecureInfo = ThreeDSecureInfo(
                            liabilityShifted = info.liabilityShifted,
                            liabilityShiftPossible = info.liabilityShiftPossible,
                            wasVerified = info.wasVerified,
                            challengeRequired = false
                        )
                    )
                }
                is ThreeDSecurePaymentAuthRequest.Failure -> {
                    Log.e(TAG, "[requestPayment] 3DS auth request failure: ${paymentAuthRequest.error.message}")
                    handleError(paymentAuthRequest.error.message, paymentAuthRequest.error.localizedMessage)
                }
            }
        }
    }

    override fun onThreeDSecurePaymentAuthResult(paymentAuthResult: ThreeDSecurePaymentAuthResult) {
        Log.d(TAG, "[onThreeDSecurePaymentAuthResult] Received 3DS auth result, tokenizing...")

        threeDSecureClient.tokenize(paymentAuthResult) { threeDSecureResult ->
            Log.d(TAG, "[onThreeDSecurePaymentAuthResult] tokenize callback, type: ${threeDSecureResult::class.simpleName}")

            when (threeDSecureResult) {
                is ThreeDSecureResult.Success -> {
                    val threeDSecureNonce = threeDSecureResult.nonce
                    val info = threeDSecureNonce.threeDSecureInfo

                    Log.d(TAG, "[onThreeDSecurePaymentAuthResult] 3DS success after challenge, liabilityShiftPossible=${info.liabilityShiftPossible}, liabilityShifted=${info.liabilityShifted}, wasVerified=${info.wasVerified}")

                    val paymentMethod = args.paymentMethod as PaymentMethod.ThreeDSecure
                    handleSuccess(
                        nonce = threeDSecureNonce.string,
                        amount = paymentMethod.amount,
                        currency = paymentMethod.currency,
                        paymentType = "Card3DS",
                        threeDSecureInfo = ThreeDSecureInfo(
                            liabilityShifted = info.liabilityShifted,
                            liabilityShiftPossible = info.liabilityShiftPossible,
                            wasVerified = info.wasVerified,
                            challengeRequired = true
                        )
                    )
                }
                is ThreeDSecureResult.Failure -> {
                    Log.e(TAG, "[onThreeDSecurePaymentAuthResult] 3DS failure: ${threeDSecureResult.error.message}")
                    handleError(threeDSecureResult.error.message, threeDSecureResult.error.localizedMessage)
                }
                is ThreeDSecureResult.Cancel -> {
                    Log.d(TAG, "[onThreeDSecurePaymentAuthResult] 3DS cancelled by user")
                    handleCancel()
                }
            }
        }
    }

    override fun onResume(intent: Intent) {
        Log.d(TAG, "[onResume] Activity resumed")
    }

    override fun onDestroy() {
        Log.d(TAG, "[onDestroy] Cleaning up")
        threeDSecureLauncher = null
    }
}
